src/include/sof/audio/source_api.h

```diff
@@ -0,0 +1,267 @@
+	 * return number of bytes that are available for immediate use
+	 */
+	uint32_t (*get_data_available)(audio_src_hnd src_handler);
+
+	/**
+	 * Retrieves a fragment of circular data to be used by the caller (to read)
+	 * After calling get_data, the data are guaranteed to be available
+	 * for exclusive use (read only) on the caller core through cached pointer
+	 * The caller MUST take care of data circularity based on provided pointers
+	 *
+	 * Depending on implementation - there may be a way to have several receivers of the same
+	 * data, as long as the receiver respects that data are read-only and won'do anything
+	 * fancy with cache handling itself
+	 *
+	 * some implementation data may be stored in linear buffer
+	 * in that case:
+	 *  data_ptr = buffer_start
+	 *  buffer_end = data_ptr + req_size
+	 *  buffer_size = req_size
+	 *
+	 *  and the data receiver may use it as usual, rollover will simple never occur
+	 *  NOTE! the caller MUST NOT assume that pointers to start/end of the circular buffer
+	 *	  are constant. They may change between calls
+	 *
+	 * @param [in] req_size requested size of data.
+	 * @param [out] data_ptr a pointer to data will be provided there
+	 * @param [out] buffer_start pointer to circular buffer start
+	 * @param [out] buffer_end pointer to circular buffer end
+	 * @param [out] buffer_size size of circular buffer
+	 *
+	 * @retval -ENODATA if req_size is bigger than available data
+	 */
+	int (*get_data)(audio_src_hnd src_handler, uint32_t req_size, void **data_ptr,
+			void **buffer_start, void **buffer_end, uint32_t *buffer_size);
+
+	/**
+	 * Releases fragment previously obtained by get_data()
+	 * Once called, the data are no longer available for the caller
+	 *
+	 * @param free_size amount of data that the caller declares as "never needed again"
+	 *
+	 * if free_size == 0 the source implementation MUST keep all data in memory and make them
+	 * available again at next get_data() call
+	 *
+	 * if free_size is bigger than the amount of data obtained before by get_data(), only
+	 * the amount obtained before will be freed. That means - if somebody obtained some data,
+	 * processed it and won't need it again, it may simple call put_data with free_size==MAXINT
+	 *
+	 * @return proper error code (0  on success)
+	 */
+	int (*put_data)(audio_src_hnd src_handler, uint32_t free_size);
+
+	/**
+	 * OPTIONAL: Notification to the source implementation about changes in audio format
+	 */
+	void (*on_audio_format_set)(audio_src_hnd src_handler);
+};
+
+struct source {
+	const struct source_ops *ops;
+	uint32_t locked_read_frag_size;
+	uint32_t num_of_bytes_processed;		     /** processed bytes counter */
+	struct sof_audio_stream_params *audio_stream_params; /** pointer to audio params */
+};
+
+static inline void source_init(audio_src_hnd src_handler, const struct source_ops *ops,
+			       struct sof_audio_stream_params *audio_stream_params)
+{
+	struct source *source = (struct source *)src_handler;
```

Comment:
let's not do that. Please define a proper opaque type instead of a `void *` macro